File: app/src/main/cpp/example16/TextureManager.cpp
//
// Created by Raining on 2020/6/11.
//

#include "TextureManager.h"
#include "esUtils.h"

static InternalRendererData *createData(int w, int h, bool cache) {
    auto *data = new InternalRendererData();
    data->width = w;
    data->height = h;
    data->cache = cache;

    glGenFramebuffers(1, &(data->framebuffer));
    data->texture = createRGBATexture2D(w, h);

    // framebuffer configuration
    glBindFramebuffer(GL_FRAMEBUFFER, data->framebuffer);
    glFramebufferTexture2D(GL_FRAMEBUFFER, GL_COLOR_ATTACHMENT0, GL_TEXTURE_2D, data->texture,
                           0);
    if (glCheckFramebufferStatus(GL_FRAMEBUFFER) != GL_FRAMEBUFFER_COMPLETE)
        ALOGE ("ERROR::FRAMEBUFFER:: Framebuffer is not complete!");
    glBindFramebuffer(GL_FRAMEBUFFER, 0);

    return data;
}

static void releaseData(InternalRendererData *data) {
    glDeleteFramebuffers(1, &(data->framebuffer));
    glDeleteTextures(1, &(data->texture));
}

static bool removeHistory(std::vector<InternalRendererData *> *list, RendererData *data) {
    bool result = false;
    if (data) {
        for (auto it = list->begin(); it != list->end(); ++it) {
            if (data == (*it)) {
                list->erase(it);
                result = true;
                break;
            }
        }
    }
    return result;
}

RendererData::RendererData() : framebuffer(-1), texture(-1), width(-1), height(-1) {
}

InternalRendererData::InternalRendererData() : cache(false) {
}

TextureManager::TextureManager() : buffer(10), leakHistory(10) {
}

TextureManager::~TextureManager() {
    for (auto const &value: buffer) {
        if (value) {
            releaseData(value);
            delete value;
        }
    }
    buffer.clear();

    for (auto const &value2: leakHistory) {
        if (value2) {
            releaseData(value2);
            delete value2;
        }
    }
    leakHistory.clear();
}

RendererData *TextureManager::getData(int w, int h, bool cache) {
    InternalRendererData *result = nullptr;
    for (auto it = buffer.begin(); it != buffer.end(); ++it) {
        result = *it;
        if (result) {
            if (result->width == w && result->height == h) {
                result->cache = cache;
                buffer.erase(it);
                leakHistory.push_back(result);
                return result;
            }
        }
    }
    result = createData(w, h, cache);
    leakHistory.push_back(result);
    return result;
}

void TextureManager::release(RendererData **data) {
    if (removeHistory(&leakHistory, *data)) {
        auto *temp = (InternalRendererData *) *data;
        if (temp->cache) {
            buffer.push_back(temp);
        } else {
            releaseData(temp);
            delete temp;
        }
        *data = nullptr;
    }
}

File: app/src/main/java/com/ideacarry/utils/TextureManager.java
package com.ideacarry.utils;

import android.opengl.GLES30;

import java.util.ArrayList;

/**
 * Created by Raining on 2020/9/2.
 */
public class TextureManager {
    private ArrayList<InternalRendererData> buffer = new ArrayList<>();
    private ArrayList<InternalRendererData> leakHistory = new ArrayList<>();

    public static class RendererData {
        public int framebuffer = -1;
        public int texture = -1;
        public int width;
        public int height;
    }

    public static class InternalRendererData extends RendererData {
        public Boolean cache;
    }

    public static InternalRendererData createData(int w, int h, boolean cache) {
        InternalRendererData result = new InternalRendererData();
        int[] params = GLUtils.crateFrameBuffer(w, h);
        result.texture = params[0];
        result.framebuffer = params[1];
        result.width = w;
        result.height = h;
        result.cache = cache;
        return result;
    }

    public static void releaseData(InternalRendererData data) {
        if (data != null) {
            if (data.framebuffer > -1) {
                GLES30.glDeleteFramebuffers(1, new int[]{data.framebuffer}, 0);
                data.framebuffer = -1;
            }
            if (data.texture > -1) {
                GLES30.glDeleteTextures(1, new int[]{data.texture}, 0);
                data.texture = -1;
            }
        }
    }

    public RendererData getData(int w, int h, boolean cache) {
        InternalRendererData result;
        int len = buffer.size();
        for (int i = 0; i < len; i++) {
            result = buffer.get(i);
            if (result != null) {
                if (result.width == w && result.height == h) {
                    result.cache = cache;
                    buffer.remove(i);
                    leakHistory.add(result);
                    return result;
                }
            }
        }
        result = createData(w, h, cache);
        leakHistory.add(result);
        return result;
    }

    public static boolean removeHistory(ArrayList<InternalRendererData> list, RendererData data) {
        boolean result = false;
        if (data != null) {
            int len = list.size();
            InternalRendererData temp;
            for (int i = 0; i < len; i++) {
                temp = list.get(i);
                if (temp == data) {
                    list.remove(i);
                    result = true;
                    break;
                }
            }
        }
        return result;
    }

    public void release(RendererData item) {
        if (item != null && removeHistory(leakHistory, item)) {
            InternalRendererData temp = (InternalRendererData) item;
            if (temp.cache) {
                buffer.add(temp);
            } else {
                releaseData(temp);
            }
        }
    }

    public void clear() {
        for (InternalRendererData value : buffer) {
            if (value != null) {
                releaseData(value);
            }
        }
        buffer.clear();

        for (InternalRendererData value2 : leakHistory) {
            if (value2 != null) {
                releaseData(value2);
            }
        }
        leakHistory.clear();
    }
}
